### PR TITLE
feat: add lazily loaded web vitals collection

### PR DESCRIPTION
**Pull request by pauldambra**
if we gather web vitals then we could do all kinds of magic

this

* adds chrome's web vitals library
* buffer its metrics events
* sends them as a $web_vitals_data event when they're all received
* or after 8 seconds if we don't get a full set

each event has the metric value `$web_vitals_blah_value` as a property
and the whole metric bag o' props `$web_vitals_blah_event` as another property

### CODE REVIEW

File: src/constants.ts [hunk 1]
```diff
@@ -12,6 +12,7 @@ export const CAMPAIGN_IDS_KEY = '__cmpns'
 export const EVENT_TIMERS_KEY = '__timers'
 export const AUTOCAPTURE_DISABLED_SERVER_SIDE = '$autocapture_disabled_server_side'
 export const HEATMAPS_ENABLED_SERVER_SIDE = '$heatmaps_enabled_server_side'
+export const WEB_VITALS_ENABLED_SERVER_SIDE = '$wvess'
```
Comment (daibhin) [inline src/constants.ts:15]: Is there a reason to truncate this? Feels less explicit that the other flags

File: src/types.ts [hunk 2]
```diff
@@ -69,6 +69,11 @@ export interface BootstrapConfig {
     featureFlagPayloads?: Record<string, JsonType>
 }
 
+export interface PerformanceCaptureConfig {
+    replay?: boolean
```
Comment (daibhin) [inline src/types.ts:73]: What does replay performance include? I'm guessing it's network performance. Maybe it could have a more explicit name to that effect

File: src/constants.ts (hunk 1 above)
Comment (pauldambra) [reply src/constants.ts:15]: only bundle size... it's not for humans so 🤷

File: src/types.ts (hunk 2 above)
Comment (pauldambra) [reply src/types.ts]: renamed this which ended up forcing m to write tests to improve the existing behaviour anyway

File: src/types.ts [hunk 3]
```diff
@@ -275,7 +283,15 @@ export interface DecideResponse {
...
-    capturePerformance?: boolean
+    /**
+     *     originally capturePerformance was replay only and so boolean true
+     *     is equivalent to { network_timing: true }
+     *     now capture performance can be separately enabled within replay
+     *     and as a standalone web vitals tracker
+     *     people can have them enabled separately
+     *     they work standalone but enhance each other
+     */
+    capturePerformance?: boolean | PerformanceCaptureConfig
```
Comment (pauldambra) [inline src/types.ts:294]: it makes more sense for `capturePerformance: true` to mean `{ network_timing: true, web_vitals: true}`

but then we're turning this on for anyone that has `capturePerformance: true` (should not be many people 🤷)

but I decided to not be too YOLO here

**Review by daibhin — APPROVED** [review]

File: src/extensions/web-vitals/index.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            $window_id: sessionIds.windowId,
+            timestamp: Date.now(),
+        })
+    }
+
+    private startCapturing = () => {
+        const { onLCP, onCLS, onFCP, onINP } = assignableWindow.postHogWebVitalsCallbacks
+
+        // register performance observers
+        const captureMetric = (metric: any) => this._capture(metric)
```
Comment (robbie-c) [inline src/extensions/web-vitals/index.ts:108]: If you do
```typescript
private _capture = (metric: any) => {
```
above, then you can just pass `this._capture` into the functions below

Comment (pauldambra) [reply src/extensions/web-vitals/index.ts:108]: iirc that this was undefined when evaluating it

but this was a week or so ago... i'll check and at least comment to help the future traveller even if we can't figure out why

Comment (robbie-c) [reply src/extensions/web-vitals/index.ts:108]: Using the arrow syntax in the definition of _capture should fix that, but yes otherwise it would be

**Comment by pauldambra** [comment]
only bumping patch since it's only "preview" and I want to test internally _and_ add remote config before launching it

**Comment by robbie-c** [comment]
Removed the stale label, what's the status of this? (No particular rush on my side, just wondering)

**Comment by pauldambra** [comment]
related to https://github.com/PostHog/posthog/issues/23045

File: src/__tests__/extensions/replay/sessionrecording.test.ts [hunk 5]
```diff
@@ -222,26 +229,71 @@ describe('SessionRecording', () => {
...
-            posthog.persistence?.register({ [CONSOLE_LOG_RECORDING_ENABLED_SERVER_SIDE]: false })
-            posthog.config.enable_recording_console_log = true
-            expect(sessionRecording['isConsoleLogCaptureEnabled']).toBe(true)
-        })
-
-        it('uses server side setting if client side setting is not set', () => {
-            posthog.config.enable_recording_console_log = undefined
-            posthog.persistence?.register({ [CONSOLE_LOG_RECORDING_ENABLED_SERVER_SIDE]: false })
-            expect(sessionRecording['isConsoleLogCaptureEnabled']).toBe(false)
+        it.each([
```
Comment (pauldambra) [inline src/__tests__/extensions/replay/sessionrecording.test.ts:232]: this should maybe be in a different PR... behaviour wasn't right here

File: src/extensions/web-vitals/index.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        if (urlHasChanged) {
+            // we need to send what we have
+            this._flushToCapture()
+            // poor performance is >4s, we wait twice that time to send
+            // this is in case we haven't received all metrics
+            // we'll at least gather some
+            this._delayedFlushTimer = setTimeout(this._flushToCapture, FLUSH_TO_CAPTURE_TIMEOUT_MILLISECONDS)
+        }
+
+        this.buffer.url = isUndefined(this.buffer.url) ? $currentUrl : $currentUrl
```
Comment (daibhin) [inline src/extensions/web-vitals/index.ts:131]: This will always return $currentUrl

Comment (pauldambra) [reply src/extensions/web-vitals/index.ts:131]:
```suggestion
	if (isUndefined(this.buffer.url)) {
		this.buffer.url = $currentUrl
	}
```

**Review by daibhin — APPROVED** [review]
